File: 05-Math-Binary_Search-Two_Pointer-Hash/choiyn33/D-27932.py
import sys

n, k = map(int, input().split())
people = list(map(int, sys.stdin.readline().split()))
people.append(people[n-1])

result = sys.maxsize


def search(low, high):
    global result
    if low > high:
        return

    mid = (low+high) // 2
    count = check(mid)

    if k < count:
        search(mid+1, high)

    else:   # 지친 사람(count)가 k명 이하일 때
        result = min(result, mid)
        search(low, mid-1)


def check(h):
    count = 0

    for i in range(n):
        me = people[i]

        if i == 0:
            after = people[i+1]

            if abs(after - me) > h:
                count += 1

        else:
            before = people[i-1]
            after = people[i+1]

            if abs(me - before) > h or abs(after - me) > h:
                count += 1

    return count


search(0, result)
print(result)

File: 05-Math-Binary_Search-Two_Pointer-Hash/06_ll/D-27932.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <climits>

using namespace std;

int N;
int K;
vector<int> arr;
int result = INT_MAX;

void binarySearch(int low, int high);

int check(int H) {
    int count = 0;

    for (int i = 0; i < N; i++) {
        if (i == 0) {
            int me = arr[i];
            int after = arr[i + 1];

            if (abs(me - after) > H) {
                count++;
            }
        }
        else {
            int before = arr[i - 1];
            int me = arr[i];
            int after = arr[i + 1];

            if (abs(me - before) > H || abs(me - after) > H) {
                count++;
            }
        }
    }

    return count;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    cin >> N >> K;
    arr.resize(N + 1);

    for (int i = 0; i < N; i++) {
        cin >> arr[i];
    }
    arr[N] = arr[N - 1];

    binarySearch(0, INT_MAX);

    cout << result << endl;

    return 0;
}

void binarySearch(int low, int high) {
    if (low > high) {
        return;
    }

    int mid = low + (high - low) / 2;
    int count = check(mid);

    if (K < count) {
        binarySearch(mid + 1, high);
    }
    else {
        result = min(result, mid);
        binarySearch(low, mid - 1);
    }
}
